### PR TITLE
Add Cluster resource template

### PR DESCRIPTION
**Pull request by nprokopic**
Note: WIP, initially based on Cluster resource from cluster-aws.

### What does this PR do?

(Please set a descriptive PR title. Use this space for additional explanations.)

### What is the effect of this change to users?

Cluster resource is deployed.

### How does it look like?

```
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  annotations:
    cluster.giantswarm.io/description: "Awesome Giant Swarm cluster"
  labels:
    app: "cluster"
    app.kubernetes.io/managed-by: "Helm"
    app.kubernetes.io/version: "0.0.0-dev"
    application.giantswarm.io/team: "turtles"
    giantswarm.io/cluster: "awesome"
    giantswarm.io/organization: "giantswarm"
    giantswarm.io/service-priority: highest
    cluster.x-k8s.io/cluster-name: "awesome"
    cluster.x-k8s.io/watch-filter: capi
    helm.sh/chart: "cluster-0.0.0-dev"
    cluster-apps-operator.giantswarm.io/watching: ""
  name: awesome
  namespace: default
spec:
  clusterNetwork:
    services:
      cidrBlocks:
        - 172.31.0.0/16
    pods:
      cidrBlocks:
        - 100.64.0.0/12
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: KubeadmControlPlane
    name: awesome
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AWSCluster
    name: awesome
```

### Any background context you can provide?

Working towards a PoC for restructured cluster apps (more details [here](https://github.com/giantswarm/giantswarm/issues/27020)).

### What is needed from the reviewers?

PTAL :)

### Do the docs need to be updated?

Not yet.

### Should this change be mentioned in the release notes?

- [x] CHANGELOG.md has been updated (if it exists)

### CODE REVIEW

**Comment by nprokopic** [comment]
It's expected that JSON schema validation is failing here, because the current validation has rules for existing cluster apps, and those rules do not work in the cluster chart.

File: helm/cluster/templates/cluster-api/cluster.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    {{- if .Values.metadata.labels }}
+    {{- range $key, $val := .Values.metadata.labels }}
+    {{ $key }}: {{ $val | quote }}
+    {{- end }}
+    {{- end }}
+  name: {{ include "resource.default.name" $ }}
+  namespace: {{ .Release.Namespace }}
+spec:
+  clusterNetwork:
+    services:
```
Comment (njuettner) [inline helm/cluster/templates/cluster-api/cluster.yaml:20]: Do we wanna sort it similar to the CR's?
That would mean starting with pods and then services in `clusterNetwork`

File: helm/cluster/templates/cluster-api/cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      cidrBlocks:
+      {{- toYaml .Values.connectivity.network.services.cidrBlocks | nindent 8 }}
+    pods:
+      cidrBlocks:
+      {{- toYaml .Values.connectivity.network.pods.cidrBlocks | nindent 8 }}
+  controlPlaneRef:
```
Comment (njuettner) [inline helm/cluster/templates/cluster-api/cluster.yaml:26]: Inside `controlPlaneRef` and `infrastructureRef` we do have the field `namespace` which we could define already

**Review by njuettner — APPROVED** [review]

File: helm/cluster/templates/cluster-api/cluster.yaml (hunk 1 above)
Comment (nprokopic) [reply helm/cluster/templates/cluster-api/cluster.yaml:20]: Sure :+1:

File: helm/cluster/templates/cluster-api/cluster.yaml (hunk 2 above)
Comment (nprokopic) [reply helm/cluster/templates/cluster-api/cluster.yaml:26]: We don't set namespace in these `*Ref` fields, as then it uses the same namespace as the resource which has the reference (`Cluster` here).

I think there is/was even a discussion in upstream CAPI to change the reference type to something like `LocalReference` so you cannot even set the namespace even if you wanted.

File: helm/cluster/values.yaml [hunk 3]
```diff
@@ -1,3 +1,27 @@
...
-{}
+connectivity:
+  network:
+    pods:
+      cidrBlocks:
+        - 100.64.0.0/12
+    services:
+      cidrBlocks:
+        - 172.31.0.0/16
+internal:
```
Comment (nprokopic) [inline helm/cluster/values.yaml:11]: We would statically set these from the `cluster-$provider` app in its Helm values like this:
```
cluster:
  internal:
    ...
```

and these would be treated as an internal API, as in internal in our apps, where the `cluster` chart exposes it to the `cluster-$provider` , but it's not meant to be set by end users. This is roughly aligned with the current behavior of `Values.internal`.

File: helm/cluster/templates/cluster-api/cluster.yaml (hunk 1 above)
Comment (njuettner) [reply helm/cluster/templates/cluster-api/cluster.yaml:20]: Agree

File: helm/cluster/templates/cluster-api/cluster.yaml (hunk 2 above)
Comment (njuettner) [reply helm/cluster/templates/cluster-api/cluster.yaml:26]: I see that make sense

**Pull request merged**